lib/code_writer.rb

```diff
@@ -7,29 +7,50 @@ def initialize(out)
   SEGMENT_TO_SYMBOL_HASH = {
     "argument" => "ARG",
     "local" => "LCL",
+    "this" => "THIS",
+    "that" => "THAT",
   }
 
   def write_push_pop(command, segment, index)
     if command == :C_POP
-      @out.puts <<~EOF
-        @SP
-        AM=M-1
-        D=M
-        @R13
-        M=D
-        @#{index}
-        D=A
-        @#{SEGMENT_TO_SYMBOL_HASH[segment]}
-        A=M+D
-        D=A
-        @R14
-        M=D
-        @R13
-        D=M
-        @R14
-        A=M
-        M=D
-      EOF
+      if segment == "temp"
+        @out.puts <<~EOF
+          @SP
+          AM=M-1
+          D=M
+          @R13
+          M=D
+          @#{5 + index}
+          D=A
+          @R14
+          M=D
+          @R13
+          D=M
+          @R14
+          A=M
+          M=D
```

Comment:
This is extremely similar to the non-`temp` case, differing only in the part which loads the address into `D` for later storage at `R14`: here you precompute the address in Ruby and load it directly with a single A-command, whereas for all the other segments you have to load the base address from RAM and do the addition dynamically. It might make this code easier to maintain and understand if you eliminated the duplication so that it’s easier to spot which part is different.

Reply:
(This isn’t a blocker to you merging it, btw, I’m just thinking out loud.)

Reply:
Fixed!

Reply:
Excellent! 🎉